crates/napi/src/env.rs

```diff
@@ -638,6 +638,25 @@ impl Env {
     Ok(unsafe { JsFunction::from_raw_unchecked(self.0, raw_result) })
   }
 
+  /// This API retrieves the file path of the currently running JS module as a URL. For a file on
+  /// the local file system it will start with `file://`.
+  ///
+  /// TODO: Determine the encoding of the `CStr` provided.
+  ///
+  /// # Errors
+  ///
+  /// The retrieved string may be empty if the add-on loading process fails to establish the
+  /// add-on's file name.
+  #[cfg(feature = "experimental")]
+  pub fn get_module_file_name(&self, f: impl FnOnce(Result<&std::ffi::CStr>)) {
```

Comment:
Design note: because the lifetime of the string is tied to the environment, and we don't model the lifetime of the environment even via `Env`, we basically have to (ab)use closures this way in order to expose a safe API.

Reply:
maybe `'static` is safe enough for this scenario?

Reply:
I'd like to make it return `Option<'static str>` directly, and return `None` if the underlying `*const c_char` is not a valid  utf8 string.
Then expose another API to return the raw `CStr`.
So that the `get_module_file_name` is enough for most of cases, and the `get_module_file_name_raw` can be used in `filepath` contains invalidate `utf8` string.

Reply:
Or we could just make the whole problem go away by giving a `String` back. Not sure this code will ever be on a hot path.

Reply:
@Brooooooklyn:

> maybe `'static` is safe enough for this scenario?

Maybe? 🤔 That would be safe _if_ the lifetime of the entire native module were covariant with `'static`. I'm uncertain if things like threads and are guaranteed to be cleaned up, which would make this. Then again, the design of the `Env` API currently assumes this, being `Copy`, so...🤷🏻 I'll let you make the call there!

RE: @Xaeroxe's commentary about "just" allocating once here: I don't see any significant objection to the suggestion. N.B., I'd adapt the suggestion to `OsString`, given the above. I doubt a single string allocation here would be significant for the vast majority of applications, since AIUI the module file name is only computed once, and then cached for the remainder of the native module's lifetime.

Reply:
@Brooooooklyn:

RE: using `str` instead of `CStr` (also addressing the proposed `get_module_file_name{,_raw}` split): I believe that the correct “strong” type to use here would be `OsStr`, rather than `str`, actually. The `TODO` remaining in this PR that I've [called out separately](https://github.com/napi-rs/napi-rs/pull/1232#discussion_r91703249) acknowledges that there's no documented guarantee of UTF-8 contents. However, since we're checking for the `file://` protocol prefix first, I think `OsStr` would be an exact match for the semantics specified by [Node-API upstream's docs for `node_api_get_module_file_name`](https://nodejs.org/docs/latest-v18.x/api/n-api.html#n_api_node_api_get_module_file_name) (emphasis mine):

> A URL containing the **absolute path of the location from which the add-on was loaded. For a file on the local file system it will start with `file://`.** The string is null-terminated and owned by `env` and must thus not be modified or freed.

Reply:
Ah, but I see that `OsStr` has a complication...the actual encoding accepted is platform-dependent. 😝

I think relying on UTF-8 paths _as an implementation_ is definitely acceptable as an MVP, which is then immediately convertible to `OsStr`. Not all valid system paths are UTF-8, but it should cover most folks' use cases until something better is found for parsing and constructing, i.e., [`os_str_bytes`](https://docs.rs/os_str_bytes/).

Reply:
RE: `Option<_>` and the proposed `get_module_file_name{,_raw}` split: I think returning `Result<$str_type, E>` instead, where `E` can give access to the original `CStr{,ing}`, would obviate the need for a split API. If a user doesn't care what the failure is, they can call [`Result::ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) to immediately convert into an `Option<_>`.